bitcoin/src/bip143.rs

```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: CC0-1.0
+// Written by the rust-bitcoin developers.
+
+//! [BIP143] Transaction Signature Verification for Version 0 Witness Program
+//!
+//! [BIP143]: <https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki>
+//!
+
+use core::fmt;
+
+use crate::blockdata::opcodes::all::*;
+use crate::consensus::{encode, Decodable, Encodable};
+use crate::consensus::encode::WriteExt;
+use crate::io;
+use crate::prelude::*;
+use crate::script::{Builder, Script, ScriptBuf};
+
+// TODO: Add derives and implement common traits by passing through to self.0
+
+/// The scriptCode as defined in [BIP143](https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki).
+pub struct ScriptCode(ScriptBuf);
+
+impl ScriptCode {
+    /// Returns the script code used for spending a P2WPKH output if this script is a script pubkey
+    /// for a P2WPKH output.
+    pub fn new_p2wpkh(wpkh: &Script) -> Option<Self> {
+        if !wpkh.is_v0_p2wpkh() {
+            return None
+        }
+        // From BIP143: For P2WPKH witness program, the scriptCode is 0x1976a914{20-byte-pubkey-hash}88ac
+        let script = Builder::new()
+            .push_opcode(OP_PUSHBYTES_25)
+            .push_opcode(OP_DUP)               // 0x76
+            .push_opcode(OP_HASH160)           // 0xa9
+            // push_slice pushes OP_PUSHBYTES_20 (0x14)
+            .push_slice(&wpkh.as_bytes()[2..]) // The `wpkh` script is 0x00, 0x14, <pubkey_hash>
+            .push_opcode(OP_EQUALVERIFY)       // 0x88
+            .push_opcode(OP_CHECKSIG)          // 0xac
+            .into_script();
+
+        Some(ScriptCode(script))
+    }
+
+    /// Returns the script code used for spending a P2WSH output using `witness` (redeem script).
+    ///
+    /// WARNING: Does not remove `OP_CODESEPARATOR`.
```

Comment:
IIUC if you supply a script with `OP_CODESEPARATOR` this will create a money-losing script. If it's true, I want the function to panic rather than losing money.

Reply:
Good point, thanks man.

Reply:
Alright, I might sound like a noob here, please correct me if I'm wrong, but this function is for determining the scriptCode which is used to unlock coins, it has nothing to do with the locking script so it cannot create a money-losing script.

Its not trivial to determine if the witness has `OP_CODESEPARATOR` in it because one has to have knowledge of the script structure since `0xab` can appear inside other data in the witness.

Reply:
Oh, I think you're right, TBH my mind wasn't perfectly clear when I wrote that. I wonder if we should protect against creation of the address instead (where we can, maybe as a debug assert).

Regarding implementation shouldn't we just iterate over `Instructions`?

Reply:
`OP_CODESEP` is just an esoteric thing that users must know what they are doing. I think writing in doc-comment is just fine. The engineering time we waste here is not going to be worth any speculated use-cases (if any) that we can think. 

I vote that we can keep the codesep issue as a very low-priority until we have someone complain about it.

Reply:
Some other useful things we can do here is to check the witness script size < 3600 (standardness) rules and < 10000 (consensus) rule.

Reply:
If calling `instructions().find(|x| matches!(x, Instrution::Op(OP_CODESEPARATOR)))` is enough, then I really want it - low effort protection against inconvenience (since *our library* doesn't support it). If it's more than that, I'd probably ignore it.